### PR TITLE
feat(dsrepl): enable WAL-less batch writes

### PR DESCRIPTION
**Pull request by keynslug**
Fixes [EMQX-12468](https://emqx.atlassian.net/browse/EMQX-12468).

Release version: v5.8

Depends on https://github.com/emqx/erlang-rocksdb/pull/33.

## Summary

Use WAL-less write mode in `emqx_ds_storage_layer` backends, which should be safe because `ra` keeps the Raft log anyway, assuming `ra` has good correctness / durability guarantees in regards to its own WAL and log segment writes.

Safety is achieved through following mechanisms.
1. The implementation of `ra_machine` from time to time asks Ra to release _all_ log entries up to the current Ra index.
2. According to internal logic, Ra often satisfies that request by taking a snapshot at the current Ra index.
3. Taking the snapshot involves flushing _all yet unflushed_ column families to disk at the storage layer.
4. Once this is done, and snapshot is successfully taken, Ra marks the logs releasable and collects them eventually.

Open questions:
1. Do we want to be able to toggle WAL-less mode as a whole or on the storage layer level, through a set of options?
2. Do we need to introduce a safety margin, by requesting release not up to the current but to some past Raft index? So that there's some extra logs to replay during snapshot recovery. This will involve a bit more messy state tracking to make it work correctly.
3. How to decide when it's time to request releasing logs? Ideally this should happen as often as is optimal for RocksDB, e.g. by default RocksDB flushes memtables once it hits 64 MiB. For that we need to have some sort of batch accounting, but this sounds suboptimal, both in the context of Ra server process (because this will block it) and in the context of egress process (because these numbers will become part of the Raft log).
4. Correspondingly, should automatic flushing be disabled?

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [ ] Added tests for the changes
- [ ] ~~Added property-based tests for code which performs user input validation~~
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [x] For internal contributor: there is a jira ticket to track this change
- [ ] ~~Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket~~
- [x] Schema changes are backward compatible

[EMQX-12468]: https://emqx.atlassian.net/browse/EMQX-12468?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/emqx_durable_storage/src/emqx_ds_storage_reference.erl [hunk 1]
```diff
@@ -115,7 +115,7 @@ commit_batch(_ShardId, #s{db = DB, cf = CF}, Messages) ->
         end,
         Messages
     ),
-    Res = rocksdb:write_batch(DB, Batch, _WriteOptions = []),
```
Comment (ieQu1) [inline apps/emqx_durable_storage/src/emqx_ds_storage_reference.erl:118]: This should be passed as a parameter to the storage layer. Replication-less edition will use rocksdb WAL.

File: apps/emqx_utils/src/emqx_utils_stream.erl [hunk 2]
```diff
@@ -190,7 +190,7 @@ transpose_tail(S, Tail) ->
 %% @doc Make a stream by concatenating multiple streams.
 -spec chain([stream(X)]) -> stream(X).
 chain(L) ->
-    lists:foldl(fun chain/2, empty(), L).
+    lists:foldr(fun chain/2, empty(), L).
```
Comment (ieQu1) [inline apps/emqx_utils/src/emqx_utils_stream.erl:193]: What is the reason of this change?

Comment (keynslug) [reply apps/emqx_utils/src/emqx_utils_stream.erl:193]: This preserves the order of `L`, as evidenced by some of the `emqx_ds` tests. Without this change, once `add_generation` events are added into a test event stream, [this call](https://github.com/emqx/emqx/blob/a37376e3530aa10b141ffb9b0e95c0149452379d/apps/emqx_durable_storage/test/emqx_ds_test_helpers.erl#L287-L288) results in a stream which goes from newer generations to older.

Comment (ieQu1) [reply apps/emqx_utils/src/emqx_utils_stream.erl:193]: Thanks!

File: apps/emqx_durable_storage/src/emqx_ds_storage_layer.erl [hunk 3]
```diff
@@ -305,21 +330,21 @@ store_batch(Shard, Messages, Options) ->
...
-    emqx_ds:message_store_opts()
+    batch_prepare_opts()
 ) -> {ok, cooked_batch()} | ignore | emqx_ds:error(_).
-prepare_batch(Shard, Messages = [{Time, _} | _], Options) ->
+prepare_batch(Shard, Messages = [{Time, _} | _], _Options) ->
     %% NOTE
     %% We assume that batches do not span generations. Callers should enforce this.
-    ?tp(emqx_ds_storage_layer_prepare_batch, #{
-        shard => Shard, messages => Messages, options => Options
-    }),
```
Comment (ieQu1) [inline apps/emqx_durable_storage/src/emqx_ds_storage_layer.erl:315]: ?

Comment (keynslug) [reply apps/emqx_durable_storage/src/emqx_ds_storage_layer.erl:315]: This was almost exactly the same tracepoint as `emqx_ds_storage_layer_store_batch`. I transformed it into `emqx_ds_storage_layer_batch_cooked` with some extra content.

File: apps/emqx_durable_storage/src/emqx_ds_storage_layer.erl [hunk 4]
```diff
@@ -305,21 +330,21 @@ store_batch(Shard, Messages, Options) ->
...
     %% FIXME: always store messages in the current generation
     case generation_at(Shard, Time) of
         {GenId, #{module := Mod, data := GenData}} ->
             T0 = erlang:monotonic_time(microsecond),
             Result =
-                case Mod:prepare_batch(Shard, GenData, Messages, Options) of
+                case Mod:prepare_batch(Shard, GenData, Messages) of
```
[the remark this answers is not in the document]
Comment (keynslug) [reply apps/emqx_durable_storage/src/emqx_ds_storage_layer.erl:343]: There's currently no options that can reasonably affect "cooking" but can affect writing itself (unless obviously we want to bake the write options into the cooked batch). So `Options` "moved" into the `commit` callback.

File: apps/emqx_durable_storage/src/emqx_ds_storage_layer.erl (hunk 3 above)
Comment (ieQu1) [reply apps/emqx_durable_storage/src/emqx_ds_storage_layer.erl:315]: I see, thanks. Now `store_batch` is merely a convenience wrapper that combines two separate primitives together, so I think it's better to remove `emqx_ds_storage_layer_store_batch` tracepoint and keep this one.

File: apps/emqx_durable_storage/src/emqx_ds_storage_layer.erl (hunk 4 above)
[the remark this answers is not in the document]
Comment (ieQu1) [reply apps/emqx_durable_storage/src/emqx_ds_storage_layer.erl:343]: I think it's worth to reserve the possibility of passing options to both callbacks, unless there is a logical or consistency-related reason not to allow `prepare_batch` to access option.

**Review by ieQu1 — APPROVED** [review]

**Comment by emqxqa** [comment]
[TestExecution](https://emqx.atlassian.net/browse/QA-3405)
